domain-mapping/sso/DM_SSO_Cookie.php

```diff
@@ -0,0 +1,195 @@
+ * then the token is used to create an session cookie to login the user.
+ */
+defined( 'ABSPATH' ) or die();
+
+class DM_SSO_Cookie {
+    /**
+     * Constructor.
+     */
+    public function __construct() {
+        add_action( 'admin_post_dark_matter_dmsso', array( $this, 'login_token' ) );
+        add_action( 'admin_post_nopriv_dark_matter_dmcheck', array( $this, 'logout_token' ) );
+        add_action( 'wp_head', array( $this, 'head_script' ) );
+        add_action( 'plugins_loaded', array( $this, 'validate_token' ) );
+    }
+
+    /**
+     * Create the JavaScript output include for logging a user in to the admin
+     * when on the Mapped domain. This is ultimately what makes the Admin Bar
+     * appear.
+     *
+     * @return void
+     */
+    public function login_token() {
+        header( 'Content-Type: text/javascript' );
+
+        /**
+         * Ensure that the JavaScript is never empty.
+         */
+        echo "// dm_sso" . PHP_EOL;
+
+        if ( is_user_logged_in() ) {
+            /**
+             * Construct an authentication token which is passed back along with an
+             * action flag to tell the front end to
+             */
+            $url = add_query_arg( array(
+                '__dm_action' => 'authorise',
+                'auth' => wp_generate_auth_cookie( get_current_user_id(), time() + ( 2 * MINUTE_IN_SECONDS ) )
+            ), $_SERVER['HTTP_REFERER'] );
+
+            printf( 'window.location.replace( "%1$s" );', esc_url_raw( $url ) );
+        }
+
+        die();
+    }
+
+    /**
+     * Create the JavaScript output for handling the logout functionality.
+     * Without this, users can end up in a state where they are logged out of
+     * the Admin domain but remain perpetually logged in to the Mapped domains.
+     *
+     * @return void
+     */
+    public function logout_token() {
+        header( 'Content-Type: text/javascript' );
+
+        /**
+         * Ensure that the JavaScript is never empty.
+         */
+        echo "// dm_sso" . PHP_EOL;
+
+        if ( false === is_user_logged_in() ) {
+            $url = add_query_arg( array(
+                '__dm_action' => 'logout'
+            ), $_SERVER['HTTP_REFERER'] );
+            printf( 'window.location.replace( "%1$s" );', esc_url_raw( $url ) );
+        }
+
+        die();
+    }
+
+    /**
+     * Adds the <script> tag which references the admin action(s) for handling
+     * cross-domain login and logout.
+     *
+     * @return void
+     */
+    public function head_script() {
+        if ( is_main_site() ) {
+            return;
+        }
+
+        /**
+         * Determine if the mapped domain is HTTPS and if so, ensure that the
+         * Admin domain is also HTTPS. If it isn't, then we cannot do the third
+         * party cookie authentication due to the differing protocols.
+         */
+        if ( is_ssl() && ( ! defined( 'FORCE_SSL_ADMIN' ) || ! FORCE_SSL_ADMIN ) ) {
```

Comment:
The execution cost of `is_ssl()` is more than the other check so should go last.